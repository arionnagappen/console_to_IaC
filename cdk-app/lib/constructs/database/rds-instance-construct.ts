import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as kms from 'aws-cdk-lib/aws-kms';

interface DBProps {
  vpc: ec2.IVpc;
  dbSubnetGroup: rds.SubnetGroup;
  dbName: string;
  dbInstanceType: ec2.InstanceType;
  dbStorage: number;
  dbEngine: rds.IInstanceEngine;
  dbCredentials: rds.Credentials;
  dbSecurityGroup: ec2.ISecurityGroup[];
  dbMultiAZ: boolean;
}

export class MyRDSInstance extends Construct {
  public readonly rdsInstance: rds.DatabaseInstance

  constructor(scope: Construct, id: string, props: DBProps){
    super(scope, id)

    const encryptionKey = new kms.Key(this, 'RdsKmsKey', {
      enableKeyRotation: true,
      alias: 'alias/myRdsKey'
    })


    this.rdsInstance = new rds.DatabaseInstance(this, 'MyRDSInstance', {
      vpc: props.vpc,
      subnetGroup: props.dbSubnetGroup,
      databaseName: props.dbName,
      instanceType: props.dbInstanceType,
      allocatedStorage: props.dbStorage,
      engine: props.dbEngine,

      credentials: props.dbCredentials,
      securityGroups: props.dbSecurityGroup,
      publiclyAccessible: false,
      storageEncrypted: true,
      storageEncryptionKey: encryptionKey,

      deletionProtection: false,
      
      multiAz: props.dbMultiAZ
    })
  }
}