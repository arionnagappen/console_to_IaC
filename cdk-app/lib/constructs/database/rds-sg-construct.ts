import { Construct } from "constructs";
import * as ec2 from 'aws-cdk-lib/aws-ec2';

export interface DBSgProps {
  vpc: ec2.IVpc;
  appSecurityGroupId: ec2.ISecurityGroup["securityGroupId"]
}

export class DatabaseSecurityGroup extends Construct {
  public readonly dbSecurityGroup: ec2.SecurityGroup;

  constructor(scope: Construct, id: string, props: DBSgProps){
    super(scope, id)

    this.dbSecurityGroup = new ec2.SecurityGroup(this, 'DBSecurityGroup', {
      vpc: props.vpc,
      description: 'RDS Database Security Group',
      allowAllOutbound: false
    })

    // Only allow inbound traffic on port 3306
    this.dbSecurityGroup.addIngressRule(
      ec2.Peer.securityGroupId(props.appSecurityGroupId),
      ec2.Port.tcp(3306)
    )

    // Only allow outbound traffic on port 3306
    this.dbSecurityGroup.addEgressRule(
      ec2.Peer.securityGroupId(props.appSecurityGroupId),
      ec2.Port.tcp(3306)
    )
  }
}