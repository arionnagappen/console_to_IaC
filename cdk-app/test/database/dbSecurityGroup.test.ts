import * as cdk from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { DatabaseSecurityGroup } from '../../lib/constructs/database/rds-sg-construct';

// Assertion Test
test('Test DB Security group for  correct rules', ()=>{
  // Instantiates the App Construct
  const app = new cdk.App();

  // Instantiates the Stack Construct & uses the App Construct as a parameter
  const stack = new cdk.Stack(app, 'TestStack');

  // Adds a Vpc
  const vpc = new ec2.Vpc(stack, 'TestVpc', {
    maxAzs: 1,
  });

  const appSG = new ec2.SecurityGroup(stack, 'TestSG', {vpc});

  new DatabaseSecurityGroup(stack, 'DbSGTest', {
    vpc: vpc, 
    appSecurityGroupId: appSG.securityGroupId
  })

  const template = Template.fromStack(stack);

  template.hasResourceProperties('AWS::EC2::SecurityGroup', {

    GroupDescription: 'RDS Database Security Group',

    // Ingress Rule on Port 3306
    SecurityGroupIngress: Match.arrayWith([
      Match.objectLike ({
        IpProtocol: 'tcp',
        FromPort: 3306,
        ToPort: 3306,
        SourceSecurityGroupId: Match.anyValue()
      })
    ]),

    // Egress Rule on Port 3306
    SecurityGroupEgress: Match.arrayWith([
      Match.objectLike ({
        IpProtocol: 'tcp',
        FromPort: 3306,
        ToPort: 3306,
        DestinationSecurityGroupId: Match.anyValue()
      })
    ])
  });
});

// Snapshot Test
test('Database Security Group matches snapshot', ()=>{
  // Instantiates the App Construct
  const app = new cdk.App();

  // Instantiates the Stack Construct & uses the App Construct as a parameter
  const stack = new cdk.Stack(app, 'TestStack');

  // Adds a Vpc
  const vpc = new ec2.Vpc(stack, 'TestVpc', {maxAzs: 1,});

  const appSG = new ec2.SecurityGroup(stack, 'TestSG', {vpc});

  new DatabaseSecurityGroup(stack, 'DbSGTest', {
    vpc: vpc, 
    appSecurityGroupId: appSG.securityGroupId
  })

  const template = Template.fromStack(stack).toJSON();

  expect(template).toMatchSnapshot();
});