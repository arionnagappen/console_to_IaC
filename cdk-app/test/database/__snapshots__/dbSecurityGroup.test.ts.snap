// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database Security Group matches snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DbSGTestDBSecurityGroupCF0DE582": {
      "Properties": {
        "GroupDescription": "RDS Database Security Group",
        "SecurityGroupEgress": [
          {
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "TestSG581D3391",
                      "GroupId",
                    ],
                  },
                  ":3306",
                ],
              ],
            },
            "DestinationSecurityGroupId": {
              "Fn::GetAtt": [
                "TestSG581D3391",
                "GroupId",
              ],
            },
            "FromPort": 3306,
            "IpProtocol": "tcp",
            "ToPort": 3306,
          },
        ],
        "SecurityGroupIngress": [
          {
            "Description": {
              "Fn::Join": [
                "",
                [
                  "from ",
                  {
                    "Fn::GetAtt": [
                      "TestSG581D3391",
                      "GroupId",
                    ],
                  },
                  ":3306",
                ],
              ],
            },
            "FromPort": 3306,
            "IpProtocol": "tcp",
            "SourceSecurityGroupId": {
              "Fn::GetAtt": [
                "TestSG581D3391",
                "GroupId",
              ],
            },
            "ToPort": 3306,
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestSG581D3391": {
      "Properties": {
        "GroupDescription": "TestStack/TestSG",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "TestVpcE77CE678": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcIGW9DD53F70": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcPrivateSubnet1DefaultRoute32E7B814": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "TestVpcPublicSubnet1NATGatewayA323E3EC",
        },
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet1RouteTable469B0105",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPrivateSubnet1RouteTable469B0105": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPrivateSubnet1RouteTableAssociationFFD4DFF7": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPrivateSubnet1RouteTable469B0105",
        },
        "SubnetId": {
          "Ref": "TestVpcPrivateSubnet1SubnetCC65D771",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPrivateSubnet1SubnetCC65D771": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/17",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcPublicSubnet1DefaultRoute6C0F0315": {
      "DependsOn": [
        "TestVpcVPCGWF1827B84",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet1RouteTable4CBFF871",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVpcPublicSubnet1EIP4884338C": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "TestVpcPublicSubnet1NATGatewayA323E3EC": {
      "DependsOn": [
        "TestVpcPublicSubnet1DefaultRoute6C0F0315",
        "TestVpcPublicSubnet1RouteTableAssociation7D1DECD9",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "TestVpcPublicSubnet1EIP4884338C",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet1SubnetA7DB1EDF",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "TestVpcPublicSubnet1RouteTable4CBFF871": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVpcPublicSubnet1RouteTableAssociation7D1DECD9": {
      "Properties": {
        "RouteTableId": {
          "Ref": "TestVpcPublicSubnet1RouteTable4CBFF871",
        },
        "SubnetId": {
          "Ref": "TestVpcPublicSubnet1SubnetA7DB1EDF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVpcPublicSubnet1SubnetA7DB1EDF": {
      "Properties": {
        "AvailabilityZone": {
          "Fn::Select": [
            0,
            {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/17",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "TestStack/TestVpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVpcVPCGWF1827B84": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "TestVpcIGW9DD53F70",
        },
        "VpcId": {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
